packages/flutter/lib/src/cupertino/nav_bar.dart

```diff
@@ -908,6 +930,160 @@ class _LargeTitleNavigationBarSliverDelegate
+    renderObject
+      ..navBarConstraints = constraints
+      ..maxExtent = maxExtent
+      ..textDirection = textDirection;
+  }
+}
+
+class _RenderLargeTitle extends RenderProxyBox {
+  _RenderLargeTitle({
+    required BoxConstraints navBarConstraints,
+    required double maxExtent,
+    required TextDirection textDirection,
+  }) : _navBarConstraints = navBarConstraints,
+      _maxExtent = maxExtent,
+      _textDirection = textDirection;
+
+  static const AlignmentDirectional alignment = AlignmentDirectional.bottomStart;
+
+  BoxConstraints get navBarConstraints => _navBarConstraints;
+  BoxConstraints _navBarConstraints;
+  set navBarConstraints(BoxConstraints value) {
+    if (_navBarConstraints == value) {
+      return;
+    }
+    _navBarConstraints = value;
+
+    markNeedsLayout();
+    markNeedsPaint();
+  }
+
+  double get maxExtent => _maxExtent;
+  double _maxExtent;
+  set maxExtent(double value) {
+    if (_maxExtent == value) {
+      return;
+    }
+    _maxExtent = value;
+
+    markNeedsLayout();
+    markNeedsPaint();
+  }
+
+  TextDirection get textDirection => _textDirection;
+  TextDirection _textDirection;
+  set textDirection(TextDirection value) {
+    if (_textDirection == value) {
+      return;
+    }
+    _textDirection = value;
+
+    markNeedsLayout();
+    markNeedsPaint();
+    markNeedsSemanticsUpdate();
+  }
+
+  Matrix4? transform;
+
+  Matrix4? get _effectiveTransform {
+    // Maximum scale lets us prevent the large title
+    // from getting clipped when it's width is greater than
+    // the navigation bar's max width constraint.
+    final double maxScale = clampDouble((navBarConstraints.maxWidth - _kNavBarEdgePadding) / size.width, 1.0, 1.15);
+
+    // This scale is estimated from the settings app in iOS 14.
+    // The large title scales linearly from 1.0 up to 1.15 magnification.
+    // The `navBarConstraints.maxHeight` value is the height of the nav bar,
+    // and `maxExtent` is the default large title height the nav bar snaps back to.
+    // The difference between the two heights is used to scale the title.
+    final double scale = clampDouble(1.0 + (navBarConstraints.maxHeight - maxExtent) / maxExtent *  0.12, 1.0, maxScale);
+
+    transform = Matrix4.diagonal3Values(scale, scale, 1.0);
+
+    final Alignment resolvedAlignment = alignment.resolve(textDirection);
+    final Matrix4 result = Matrix4.identity();
+
+    Offset? translation;
+    if (resolvedAlignment != null) {
+      translation = resolvedAlignment.alongSize(size);
+      result.translate(translation.dx, translation.dy);
+    }
+    result.multiply(transform!);
+    if (resolvedAlignment != null) {
+      result.translate(-translation!.dx, -translation.dy);
+    }
+
+    return result;
+  }
+
+  @override
+  void performLayout() {
+    child!.layout(constraints, parentUsesSize: true);
+    size = child!.size;
+
+    transform = _effectiveTransform;
```

Comment:
you set `transform` on line 1053 and on line 1030, is that intentional? it confuses the ownership model a bit.

i think you can remove line 1030 and just inline the Matrix4.diagonal3Values into line 1040 (the multiply)? you might be able to find a member on Matrix4 that just does the scale directly actually.

Reply:
yeah, i've just noticed .scale() on Matrix4, thanks.